### DISCUSSION TITLE
Rich querying  API for content collections

### DISCUSSION THREAD

**Discussion by snimavat** [withastro/roadmap#574] [Proposal]
### Body

# Summary

Currently the Astro content support is based on [getCollection()](https://docs.astro.build/en/reference/api-reference/#getcollection) and [getEntryBySlug()](https://docs.astro.build/en/reference/api-reference/#getentrybyslug)

This is good enough for most use cases, however it lacks rich query API.
Some similar frameworks such as [NuxtJS Content module](https://content.nuxtjs.org/guide/displaying/querying) provides rich set of API for querying content.

# Background & Motivation
Lack of rich set of querying API forces developers to implement it their self. 
Other similar frameworks are available which provides excellent querying capabilities for markdown content.

It will help developers implement complex, nested content organisation, and flexibility to query and display
based on different criteria effectively and without code duplication.

# Goals
- Capability to query by different fields of front matter
- Capability to query certain fields only instead of whole glob
- capability to query just single, or multiple documents
- Capability to use `max`, `limit`, `offset` for queries returning list of content entries
- Support for basic. criterias, such as '$eq', '$gt' '$lt', '$in' etc

# Example
- `queryContent('articles').where({ title: 'Home' }).findOne()`
- `const articles = await queryContent('articles').where({ name: { $in: ['odin', 'thor'] } }).find()`
- `queryContent('articles').skip(5).limit(5).find()`
- queryContent('articles').only('id').find()

**Comment by matthewp** [comment]
Thanks for submitting. This is something we are _very_ interested in supporting, as we have seen the limits that come with attempting to solve every use-case with specific APIs. The exact shape is unknown, but this is something we are likely working on as we approach 3.0.

**Comment by stereobooster** [comment]
Related: https://github.com/withastro/roadmap/discussions/689
For inspiration: https://github.com/blacksmithgu/obsidian-dataview
